smalltalksrc/VMMaker/StackInterpreter.class.st

```diff
@@ -15934,6 +15995,24 @@ StackInterpreter >> validStackPageBaseFrames [
 	^true
 ]
 
+{ #category : #'indexing primitive support' }
+StackInterpreter >> validateArray: array andIndex: index [
+	"This is a helper method to valide if the array and the index parameter are valid.
+	For example, that the index is an integer, that the array is not a forwarder, etc."
+
+	<inline: true>
+	"Validate that the receiver is not immediate"
+	(objectMemory isImmediate: array)
+		ifTrue: [ ^ self primitiveFailFor: PrimErrInappropriate ].
+
+	"If the index is not an integer, fail"
+	"If the array is a forwarder, for example in the case of e.g. object:basicAt:, fail"
+	(objectMemory isNonIntegerObject: index)	
+		ifTrue: [ ^ self primitiveFailFor: PrimErrBadArgument ].
+	(objectMemory isForwarded: array)
+		ifTrue: [ ^ self primitiveFailFor: PrimErrBadArgument ]
+]
+
```

Comment:
Take care ⚠️ 
- You cannot move these validations outside the main method, because the returns were prepared to cut the flow. If you want to do this you have to check `self successful` after this method is called.
- I think that validation is missing also: `argumentCount > 1 "e.g. object:basicAt:"`